include/zephyr/bluetooth/conn.h

```diff
@@ -1358,6 +1358,18 @@ struct bt_conn_auth_info_cb {
  */
 int bt_conn_auth_cb_register(const struct bt_conn_auth_cb *cb);
 
+/** @brief Register authentication callbacks used only for a given connection.
+ *
+ *  Register callbacks to handle authenticated pairing for a given connection.
+ *  The set of callbacks overlays globally defined callbacks.
+ *
```

Comment:
The `bt_conn_connected` defined in `host/conn.c` seems to call `bt_l2cap_connected` before `notify_connected`. Because of that the registered connection callbacks should be called after l2cap channels are notified about the connection.

Reply:
I updated API doc informing about the mentioned root causes of returning error code.
I would not suggest that the callbacks must be updated in the connected callbacks. This could actually be done at any point until the per-connection callbacks are already assigned or security procedures in the SMP module have already started.